worker/upload.go

```diff
@@ -0,0 +1,1154 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/montanaflynn/stats"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"go.sia.tech/renterd/tracing"
+	"lukechampine.com/frand"
+)
+
+var (
+	errNoFreeQueue    = errors.New("no free queue")
+	errNotEnoughHosts = errors.New("not enough hosts to support requested redundancy")
+)
+
+type (
+	uploadID [8]byte
+
+	uploader struct {
+		hp hostProvider
+		rl revisionLocker
+
+		overdriveTimeout time.Duration
+		maxOverdrive     uint64
+
+		statsOverdrive *dataPoints
+		statsSpeed     *dataPoints
+
+		stopChan chan struct{}
+
+		mu        sync.Mutex
+		contracts []*uploadQueue
+		completed map[uploadID]map[uploadID]map[types.FileContractID]struct{}
+		excluded  map[uploadID]map[types.FileContractID]struct{}
+		history   map[uploadID][]uploadID
+		used      map[uploadID]map[uploadID]map[types.FileContractID]struct{}
+
+		nextSlabTriggers        map[uploadID]chan struct{}
+		sectorCompletedTriggers map[uploadID]chan struct{}
```

Comment:
All the `map[uploadID]foo` maps could be a single `uploads map[uploadID]*upload` field. That way we don't end up forgetting to clear one of those maps by accident since there is only one to delete from.